<a href="https://colab.research.google.com/github/Jigyasa-sharma7/Pollution-Level-Forecasting/blob/main/pollution_forecasting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import glob
import os

Pollution Level Forecasting — Dataset Understanding
1. Project Objective

The goal of this project is to predict PM2.5 pollution levels using different environmental, weather, time, and location-related factors.

Target Variable

PM2.5 → The pollution level we want to predict.

Independent Variables / Features

We will use information such as:

Other air pollutants
Weather conditions
Wind conditions
Time of day
Date/season
Monitoring station

The project is therefore a Machine Learning Regression problem because PM2.5 is a continuous numerical value.

2. Dataset Information

The dataset contains hourly air pollution measurements from 12 monitoring stations.

The data covers approximately:

1 March 2013 → 28 February 2017

Our dataset contains approximately:

420,768 observations

Each observation represents pollution and weather conditions recorded at a particular station at a particular hour.

3. Features in the Dataset
A. Pollution Features
PM2.5 — Particulate Matter 2.5

PM2.5 stands for Particulate Matter with a diameter of 2.5 micrometers or smaller.

These are very small particles that can penetrate deep into the respiratory system.

Role in our project:

PM2.5 is our target variable.

In simple words:

We want our machine learning model to predict the PM2.5 value.

PM10 — Particulate Matter 10

PM10 stands for Particulate Matter with a diameter of 10 micrometers or smaller.

PM10 particles are generally larger than PM2.5.

PM10 can be useful for predicting PM2.5 because both represent particulate pollution.

SO2 — Sulfur Dioxide

SO2 stands for Sulfur Dioxide.

It is a gas commonly produced by burning sulfur-containing fuels such as coal and oil.

Possible sources include:

Industries
Power plants
Fuel combustion

SO2 can therefore provide information about pollution from combustion and industrial activities.

NO2 — Nitrogen Dioxide

NO2 stands for Nitrogen Dioxide.

It is commonly produced by combustion processes.

Important sources include:

Vehicle exhaust
Industrial combustion
Power generation

NO2 is particularly useful when studying traffic-related air pollution.

CO — Carbon Monoxide

CO stands for Carbon Monoxide.

It is a colorless and odorless gas mainly produced by incomplete combustion.

Possible sources include:

Vehicle exhaust
Burning fuels
Industrial processes
O3 — Ozone

O3 stands for Ozone.

In this dataset, it refers to ground-level ozone, which is an air pollutant.

Ground-level ozone can form through chemical reactions involving other pollutants, especially in the presence of sunlight.

4. Weather / Meteorological Features
TEMP — Temperature

TEMP represents the air temperature.

The values are approximately measured in:

°C (degrees Celsius)

Temperature can affect:

Atmospheric conditions
Chemical reactions
Pollution formation
Pollution dispersion
PRES — Atmospheric Pressure

PRES represents atmospheric pressure.

It is approximately measured in:

hPa (hectopascals)

Atmospheric pressure can influence weather conditions and the movement or accumulation of pollutants.

DEWP — Dew Point Temperature

DEWP stands for Dew Point Temperature.

The dew point is the temperature at which air becomes saturated with moisture.

In simple terms:

Dew point gives us information about the amount of moisture in the air.

Higher dew point generally means the air contains more moisture.

RAIN — Rainfall

RAIN represents the amount of rainfall/precipitation recorded during the hour.

Rain can remove particles from the atmosphere.

Therefore:

Rainfall can influence PM2.5 concentration.

WSPM — Wind Speed

WSPM represents wind speed.

It is measured in approximately:

meters per second (m/s)

Wind speed is very important for pollution forecasting.

For example:

Low wind speed
      ↓
Pollutants may remain concentrated
      ↓
Higher pollution levels

Whereas:

High wind speed
      ↓
Pollutants can disperse more easily
      ↓
Pollution may decrease
5. Wind Direction
wd — Wind Direction

wd stands for Wind Direction.

It tells us the direction from which the wind is coming.
We currently have 16 directions. N,E,S,W(north,east,south,west)and combination of these
6. Time Features
year

Represents the year of the observation.

Examples:

2013
2014
2015
2016
2017
month

Represents the month.

1  = January
2  = February
3  = March
...
12 = December

Month can help the model learn seasonal pollution patterns.

day

Represents the day of the month.

Possible values:

1 → 31
hour

Represents the hour of the day.

Possible values:

0 → 23

This is particularly important for pollution forecasting because pollution can vary according to the time of day.

For example:

Morning
   ↓
Traffic increases
   ↓
Pollution may increase

Similarly, evening traffic can also affect pollution levels.

7. datetime

The original dataset provides:

year
month
day
hour

We combined these into a new feature:

datetime

For example:

year  = 2013
month = 3
day   = 1
hour  = 0

becomes:

2013-03-01 00:00:00

Our dataset covers:

Earliest:
2013-03-01 00:00:00

Latest:
2017-02-28 23:00:00

The datetime column is useful for:

Time-series analysis
Sorting observations chronologically
Creating time-based features
Creating lag features
Forecasting future pollution
8. station

station represents the air-quality monitoring station where the observation was recorded.

Our dataset contains:

12 different monitoring stations.

Different stations can have different pollution patterns because their surroundings can differ.

For example:

Industrial area
      ↓
Different pollution pattern

Residential area
      ↓
Different pollution pattern

Traffic-heavy area
      ↓
Different pollution pattern

Therefore, we should keep station as an important feature.

9. No

No is basically the record/row number.

It does not represent an environmental condition.

Therefore, it will most likely be removed before machine learning because it doesn't provide meaningful information for predicting PM2.5.

In [2]:
files = glob.glob("/content/*.csv")

print("Number of CSV files:", len(files))

for file in files:
    print(os.path.basename(file))

Number of CSV files: 12
PRSA_Data_Dongsi_20130301-20170228.csv
PRSA_Data_Tiantan_20130301-20170228.csv
PRSA_Data_Changping_20130301-20170228.csv
PRSA_Data_Guanyuan_20130301-20170228.csv
PRSA_Data_Gucheng_20130301-20170228.csv
PRSA_Data_Wanliu_20130301-20170228.csv
PRSA_Data_Shunyi_20130301-20170228.csv
PRSA_Data_Dingling_20130301-20170228.csv
PRSA_Data_Aotizhongxin_20130301-20170228.csv
PRSA_Data_Nongzhanguan_20130301-20170228.csv
PRSA_Data_Huairou_20130301-20170228.csv
PRSA_Data_Wanshouxigong_20130301-20170228.csv


In [4]:
df_list = []

for file in files:
    temp_df = pd.read_csv(file)
    df_list.append(temp_df)

df = pd.concat(df_list, ignore_index=True)

print("Dataset shape:", df.shape)

Dataset shape: (420768, 18)


In [5]:
df.head()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,9.0,9.0,3.0,17.0,300.0,89.0,-0.5,1024.5,-21.4,0.0,NNW,5.7,Dongsi
1,2,2013,3,1,1,4.0,4.0,3.0,16.0,300.0,88.0,-0.7,1025.1,-22.1,0.0,NW,3.9,Dongsi
2,3,2013,3,1,2,7.0,7.0,NaN,17.0,300.0,60.0,-1.2,1025.3,-24.6,0.0,NNW,5.3,Dongsi
3,4,2013,3,1,3,3.0,3.0,5.0,18.0,NaN,NaN,-1.4,1026.2,-25.5,0.0,N,4.9,Dongsi
4,5,2013,3,1,4,3.0,3.0,7.0,NaN,200.0,84.0,-1.9,1027.1,-24.5,0.0,NNW,3.2,Dongsi


In [6]:
df.columns.tolist()

['No',
 'year',
 'month',
 'day',
 'hour',
 'PM2.5',
 'PM10',
 'SO2',
 'NO2',
 'CO',
 'O3',
 'TEMP',
 'PRES',
 'DEWP',
 'RAIN',
 'wd',
 'WSPM',
 'station']

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420768 entries, 0 to 420767
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       420768 non-null  int64  
 1   year     420768 non-null  int64  
 2   month    420768 non-null  int64  
 3   day      420768 non-null  int64  
 4   hour     420768 non-null  int64  
 5   PM2.5    412029 non-null  float64
 6   PM10     414319 non-null  float64
 7   SO2      411747 non-null  float64
 8   NO2      408652 non-null  float64
 9   CO       400067 non-null  float64
 10  O3       407491 non-null  float64
 11  TEMP     420370 non-null  float64
 12  PRES     420375 non-null  float64
 13  DEWP     420365 non-null  float64
 14  RAIN     420378 non-null  float64
 15  wd       418946 non-null  object 
 16  WSPM     420450 non-null  float64
 17  station  420768 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 57.8+ MB


In [8]:
print(df.shape)
print(df.columns.tolist())

(420768, 18)
['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2', 'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station']


In [9]:
missing = df.isnull().sum()

print(missing)

No             0
year           0
month          0
day            0
hour           0
PM2.5       8739
PM10        6449
SO2         9021
NO2        12116
CO         20701
O3         13277
TEMP         398
PRES         393
DEWP         403
RAIN         390
wd          1822
WSPM         318
station        0
dtype: int64


In [10]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_table = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": missing_percentage
})

missing_table.sort_values("Missing Percentage", ascending=False)

,Missing Values,Missing Percentage
CO,20701,4.919813
O3,13277,3.155421
NO2,12116,2.879497
SO2,9021,2.143937
PM2.5,8739,2.076916
PM10,6449,1.532674
wd,1822,0.433018
DEWP,403,0.095777
TEMP,398,0.094589
PRES,393,0.093401


In [11]:
print("Number of stations:", df["station"].nunique())
print("\nStations:")
print(df["station"].unique())

Number of stations: 12

Stations:
['Dongsi' 'Tiantan' 'Changping' 'Guanyuan' 'Gucheng' 'Wanliu' 'Shunyi'
 'Dingling' 'Aotizhongxin' 'Nongzhanguan' 'Huairou' 'Wanshouxigong']


In [12]:
print(df["PM2.5"].describe())
print("Missing PM2.5:", df["PM2.5"].isnull().sum())

count    412029.000000
mean         79.793428
std          80.822391
min           2.000000
25%          20.000000
50%          55.000000
75%         111.000000
max         999.000000
Name: PM2.5, dtype: float64
Missing PM2.5: 8739


We check the 999 value is actually a max value or an invalid value

In [13]:
print("PM2.5 values equal to 999:", (df["PM2.5"] == 999).sum())
print("PM10 values equal to 999:", (df["PM10"] == 999).sum())
print("CO values equal to 999:", (df["CO"] == 999).sum())
print("NO2 values equal to 999:", (df["NO2"] == 999).sum())
print("SO2 values equal to 999:", (df["SO2"] == 999).sum())
print("O3 values equal to 999:", (df["O3"] == 999).sum())

PM2.5 values equal to 999: 1
PM10 values equal to 999: 3
CO values equal to 999: 0
NO2 values equal to 999: 0
SO2 values equal to 999: 0
O3 values equal to 999: 0


In [14]:
df[df["PM2.5"] == 999][["station", "year", "month", "day", "hour", "PM2.5"]].head(20)

,station,year,month,day,hour,PM2.5
411482,Wanshouxigong,2016,2,8,2,999.0


We will replace the 999 values as we can conclude that they are invalid values and not a maximumvalue of our target variable

In [15]:
pollution_cols = ["PM2.5", "PM10", "CO", "NO2", "SO2", "O3"]

for col in pollution_cols:
    df[col] = df[col].replace(999, pd.NA)

In [16]:
for col in pollution_cols:
    print(f"{col} missing values:", df[col].isna().sum())

PM2.5 missing values: 8740
PM10 missing values: 6452
CO missing values: 20701
NO2 missing values: 12116
SO2 missing values: 9021
O3 missing values: 13277


We will not replace the missing values of the target variable with mean or median as it is only 2% missing values and we can delete the whole rows where PM2.5 have missing values also we have many rows so 2% reduction will not affect out model

In [17]:
df_model = df.dropna(subset=["PM2.5"]).copy()

In [18]:
print("PM2.5 = 999:", (df["PM2.5"] == 999).sum())
print("PM10 = 999:", (df["PM10"] == 999).sum())

PM2.5 = 999: 0
PM10 = 999: 0


In [19]:
missing = df.isnull().sum()
missing_percentage = (missing / len(df)) * 100

missing_table = pd.DataFrame({
    "Missing Values": missing,
    "Missing Percentage": missing_percentage
})

missing_table.sort_values("Missing Percentage", ascending=False)

,Missing Values,Missing Percentage
CO,20701,4.919813
O3,13277,3.155421
NO2,12116,2.879497
SO2,9021,2.143937
PM2.5,8740,2.077154
PM10,6452,1.533387
wd,1822,0.433018
DEWP,403,0.095777
TEMP,398,0.094589
PRES,393,0.093401


In [20]:
df["datetime"] = pd.to_datetime(
    df[["year", "month", "day", "hour"]]
)

print(df["datetime"].head())

0   2013-03-01 00:00:00
1   2013-03-01 01:00:00
2   2013-03-01 02:00:00
3   2013-03-01 03:00:00
4   2013-03-01 04:00:00
Name: datetime, dtype: datetime64[ns]


In [21]:
print("Earliest:", df["datetime"].min())
print("Latest:", df["datetime"].max())

Earliest: 2013-03-01 00:00:00
Latest: 2017-02-28 23:00:00


In [22]:
df = df.sort_values(["station", "datetime"]).reset_index(drop=True)

print(df[["station", "datetime", "PM2.5"]].head(10))

        station            datetime PM2.5
0  Aotizhongxin 2013-03-01 00:00:00   4.0
1  Aotizhongxin 2013-03-01 01:00:00   8.0
2  Aotizhongxin 2013-03-01 02:00:00   7.0
3  Aotizhongxin 2013-03-01 03:00:00   6.0
4  Aotizhongxin 2013-03-01 04:00:00   3.0
5  Aotizhongxin 2013-03-01 05:00:00   5.0
6  Aotizhongxin 2013-03-01 06:00:00   3.0
7  Aotizhongxin 2013-03-01 07:00:00   3.0
8  Aotizhongxin 2013-03-01 08:00:00   3.0
9  Aotizhongxin 2013-03-01 09:00:00   3.0


In [33]:
station_counts = df.groupby("station").size()
print("Total observations:", station_counts.sum())
print("Number of stations:", station_counts.size)

Total observations: 420768
Number of stations: 12


In [34]:
numeric_cols = [
    "CO", "O3", "NO2", "SO2", "PM10",
    "DEWP", "TEMP", "PRES", "RAIN", "WSPM"
]

print(df[numeric_cols].isnull().sum())

CO       0
O3       0
NO2     22
SO2      0
PM10     0
DEWP     0
TEMP     0
PRES     0
RAIN     0
WSPM     0
dtype: int64


In [35]:
# Convert columns to numeric
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors="coerce")

What is happening here?

Suppose one station has:

Time-------TEMP
10:00------20
11:00------NaN
12:00------24

Interpolation estimates:

11:00------22

But because we're doing:

groupby("station")

the calculation is separate for each monitoring station.

In [36]:
df[numeric_cols] = (
    df.groupby("station")[numeric_cols]
      .transform(lambda group: group.interpolate(method="linear"))
)

In [37]:
print(df[numeric_cols].isnull().sum())

CO       0
O3       0
NO2     22
SO2      0
PM10     0
DEWP     0
TEMP     0
PRES     0
RAIN     0
WSPM     0
dtype: int64


Only 22 NO2 values remain missing. These are most likely values at the very beginning or end of a station's time series, where linear interpolation can't work because there isn't a value on both sides.

Let's inspect those 22 rows

In [38]:
df[df["NO2"].isna()][
    ["station", "datetime", "NO2"]
]

,station,datetime,NO2
70128,Dingling,2013-03-01 00:00:00,NaN
70129,Dingling,2013-03-01 01:00:00,NaN
175320,Gucheng,2013-03-01 00:00:00,NaN
175321,Gucheng,2013-03-01 01:00:00,NaN
175322,Gucheng,2013-03-01 02:00:00,NaN
175323,Gucheng,2013-03-01 03:00:00,NaN
175324,Gucheng,2013-03-01 04:00:00,NaN
175325,Gucheng,2013-03-01 05:00:00,NaN
175326,Gucheng,2013-03-01 06:00:00,NaN
175327,Gucheng,2013-03-01 07:00:00,NaN


Linear interpolation couldn't fill them because there is no earlier NO2 value to interpolate from. We should not delete those rows, because 22 out of 420k is tiny and those rows may still contain useful information.
Since they're only at the beginning of a station's data, we can use forward filling within each station.these 22 values are at the very beginning of their station's data, so ffill() has nothing before them to copy from.Better solution: forward fill + backward fill

In [43]:
df["NO2"] = (
    df.groupby("station")["NO2"]
      .transform(lambda x: x.ffill().bfill())
)

In [44]:
print("Remaining missing NO2:", df["NO2"].isna().sum())

Remaining missing NO2: 0


In [45]:
print(df[numeric_cols].isnull().sum())

CO      0
O3      0
NO2     0
SO2     0
PM10    0
DEWP    0
TEMP    0
PRES    0
RAIN    0
WSPM    0
dtype: int64


In [46]:
print(df["wd"].value_counts(dropna=False))

wd
NE     43335
ENE    34142
NW     32600
N      30869
E      29752
SW     28756
NNE    28232
NNW    25326
WNW    24375
ESE    24220
SSW    22563
SE     20074
WSW    20040
S      19405
SSE    17694
W      17563
NaN     1822
Name: count, dtype: int64


We don't want to use one direction for the entire dataset because different stations can have different prevailing wind patterns.So if Station A has a missing wd, we'll use Station A's most common wind direction, not the overall most common direction.

In [47]:
station_wind_mode = (
    df.groupby("station")["wd"]
      .agg(lambda x: x.mode()[0] if not x.mode().empty else "N")
)

print(station_wind_mode)

station
Aotizhongxin      NE
Changping        NNW
Dingling         NNW
Dongsi           ENE
Guanyuan          NE
Gucheng            N
Huairou           NW
Nongzhanguan     ENE
Shunyi           NNE
Tiantan          ENE
Wanliu            NE
Wanshouxigong     NE
Name: wd, dtype: object


In [48]:
df["wd"] = df["wd"].fillna(
    df["station"].map(station_wind_mode)
)

In [50]:
print("Missing wind direction:", df["wd"].isna().sum())
print(df["wd"].value_counts())

Missing wind direction: 0
wd
NE     43699
ENE    34376
NW     32902
N      31028
E      29752
SW     28756
NNE    28715
NNW    25606
WNW    24375
ESE    24220
SSW    22563
SE     20074
WSW    20040
S      19405
SSE    17694
W      17563
Name: count, dtype: int64


In [56]:
df_model = df.dropna(subset=["PM2.5"]).copy()

In [57]:
print("Original dataset:", df.shape)
print("Modeling dataset:", df_model.shape)
print("Missing PM2.5:", df_model["PM2.5"].isna().sum())

Original dataset: (420768, 19)
Modeling dataset: (412028, 19)
Missing PM2.5: 0


Final DATA Quality check

In [58]:
print("Missing values in df_model:")
print(df_model.isnull().sum())

Missing values in df_model:
No          0
year        0
month       0
day         0
hour        0
PM2.5       0
PM10        0
SO2         0
NO2         0
CO          0
O3          0
TEMP        0
PRES        0
DEWP        0
RAIN        0
wd          0
WSPM        0
station     0
datetime    0
dtype: int64


In [59]:
print("Number of duplicate rows:", df_model.duplicated().sum())

Number of duplicate rows: 0


In [60]:
print("Rows:", df_model.shape[0])
print("Columns:", df_model.shape[1])

Rows: 412028
Columns: 19


In [61]:
df_model.info()

<class 'pandas.core.frame.DataFrame'>
Index: 412028 entries, 0 to 420767
Data columns (total 19 columns):
 #   Column    Non-Null Count   Dtype         
---  ------    --------------   -----         
 0   No        412028 non-null  int64         
 1   year      412028 non-null  int64         
 2   month     412028 non-null  int64         
 3   day       412028 non-null  int64         
 4   hour      412028 non-null  int64         
 5   PM2.5     412028 non-null  object        
 6   PM10      412028 non-null  float64       
 7   SO2       412028 non-null  float64       
 8   NO2       412028 non-null  float64       
 9   CO        412028 non-null  float64       
 10  O3        412028 non-null  float64       
 11  TEMP      412028 non-null  float64       
 12  PRES      412028 non-null  float64       
 13  DEWP      412028 non-null  float64       
 14  RAIN      412028 non-null  float64       
 15  wd        412028 non-null  object        
 16  WSPM      412028 non-null  float64       
 

In [62]:
df_model["PM2.5"] = pd.to_numeric(df_model["PM2.5"], errors="coerce")

In [63]:
print(df_model["PM2.5"].dtype)

float64


In [65]:
df_model.to_csv("pollution_model_data.csv", index=False)

In [66]:
df.to_csv("pollution_cleaned_data.csv", index=False)

## Data Cleaning Summary

The dataset initially contained 420,768 observations from 12 monitoring stations.

The following preprocessing steps were performed:

1. Combined data from all 12 monitoring stations.
2. Created a `datetime` column using year, month, day, and hour.
3. Sorted observations by station and datetime.
4. Identified invalid `999` values in pollution measurements and replaced them with missing values.
5. Converted pollution and weather variables to appropriate numerical data types.
6. Filled missing numerical measurements using station-wise time-based interpolation.
7. Handled remaining boundary missing values using forward/backward filling within each station.
8. Filled missing wind-direction (`wd`) values using the most frequent wind direction for each station.
9. Removed observations where the target variable `PM2.5` was unavailable.
10. Checked for duplicate observations.

After preprocessing, the final modeling dataset contains:

- 412,028 observations
- 19 columns
- 0 missing values
- 0 duplicate rows

The target variable for the prediction task is `PM2.5`.